Explain Client and Server Auth on domain public API (#691)
digital-asset/docs.daml.com: docs/2.8.4/docs/canton/usermanual/apis.rst

   :end-before: architecture-handbook-entry-end: TlsConnect
   :dedent:

Server Authentication
^^^^^^^^^^^^^^^^^^^^^

Canton has two ways to perform server authentication to protect from man in the middle attacks. Firstly, if TLS is used on the public API as described above, TLS also takes care of server authentication as that is one of the core functions of TLS.

> Secondly, like all nodes, the sync domain has an identity which corresponds to the fingerprint of its namespace root key. It reports its identity to connecting participant nodes and signs all its messages with keys authorized by that namespace root key on the topology ledger. Assuming no key compromises, this gives participants a guarantee that the reported identity is authentic. Thus server authentication can also be performed by the sync domain operator passing their sync domain identity to the participant node operator, and checking that the identity matches that reported by the sync domain to the participant node. The domain identity of the sole connected sync domain can be read out using console commands like:

Author: I decided not to add this. It feels to me like that detail doesn't add much.

.. code-block:: none
  
    participant1.domains.list_connected.last.domainId.filterString

Client Authentication
^^^^^^^^^^^^^^^^^^^^^

> Unlike the ledger or admin API, the public API uses Canton's cryptography and topology state for client authentication rather than mTLS. To connect to the public APIs, clients need to connect in two steps:

Author: ```suggestion
Unlike the ledger or admin API, the public API uses Canton's cryptography and topology state for client authentication rather than mutual TLS (mTLS). To connect to the public APIs, clients need to connect in two steps:
```

1. Call the `SequencerConnectService` to align on Canton Protocol versions.

> 2. Call the `SequencerAuthenticationService` to authenticate and get an access token for the other services.

Author: At what stage does the process stop in a permissioned domain? Does that step of pushing minimal topology state get rejected for a permissioned domain?

> 2. Call the `SequencerAuthenticationService` to authenticate and get an access token for the other services.

Reviewer: the member will try to push its minimal topology state to the domain, but the domain will reject it if the domain is in permissioned mode and hasn't allowed that particular member id.

3. Connect to the main `SequencerService` using the access token from 1.

> During stage 2, the client claims an identity, which is the fingerprint of a namespace root key. If that identity is allowed to connect at all (see :ref:`permissioned-domains`), then the domain responds with a challenge consisting of a nonce and all known key fingerprints of keys that are allowed to authenticate for that identity as per the topology ledger. During the first connection, that would be the namespace root key alone.

Reviewer: last sentence: it's actually the (transaction) signing key of the node, not the namespace root key, that is used to sign the challenge.

.. code-block:: none
  
    participant1.domains.list_connected.last.domainId.filterString

Client Authentication
^^^^^^^^^^^^^^^^^^^^^

Unlike the ledger or admin API, the public API uses Canton's cryptography and topology state for client authentication rather than mutual TLS (mTLS). To connect to the public APIs, clients need to connect in two steps:

1. The client calls the `SequencerConnectService` to align on Canton Protocol versions and obtain the domain id.
2. During the first connection, the client registers by sending its minimal topology state (identity, key delegations, public keys) to the domain manager.
3. The client calls the `SequencerAuthenticationService` to authenticate using a challenge-response protocol and get an access token for the other sequencer services.
4. The client connects to the main `SequencerService` using the access token from 3.

> The information the client provides in step 2 is verifiable since it's a certificate chain of keys. The domain rejects this if either the namespace root key fingerprint included is not permissioned  (see :ref:`permissioned-domains`), or if the topology state provided is invalid.

Reviewer 2: ```suggestion
The information the client provides in step 2 is verifiable since it is a certificate chain of keys. The domain rejects this if the namespace root key fingerprint included is not permissioned  (see :ref:`permissioned-domains`) or if the topology state provided is invalid.
```

> During stage 3, the client claims an identity, which is the fingerprint of a namespace root key. If that identity registered (as done in step 2), then the domain responds with a challenge consisting of a nonce and all all fingerprints of signing keys authorized for that namespace as per the topology ledger.

Reviewer 2: ```suggestion
During step 3, the client claims an identity, which is the fingerprint of a namespace root key. If that identity is registered (as done in step 2), the domain responds with a challenge consisting of a nonce and all fingerprints of signing keys authorized for that namespace as per the topology ledger.
```

> If the challenge is met successfully by signing the nonce appropriately with a key matching one of the authorized keys, the SequencerAuthenticationService responds with a time limited token which can be used to authenticate more cheaply on the other public api services.

Reviewer 2: ```suggestion
If the challenge is met successfully by signing the nonce appropriately with a key matching one of the authorized keys, the SequencerAuthenticationService responds with a time-limited token which can be used to authenticate more cheaply on the other public API services.
```

> This authentication mechanism of the restricted services is built into the public sequencer api. You don't need to do anything in order to set this up; it is enforced

Reviewer 2: ```suggestion
This authentication mechanism for the restricted services is built into the public sequencer API. You don't need to do anything to set this up; it is enforced
```

automatically and can't be turned off.

> The expiry of the token generated in step 2 during the handshake between the node and the sequencer can be configured. By default, it is valid for one hour.

Reviewer 2: ```suggestion
The expiration of the token generated in step 2 is valid for one hour by default.
```
